### DISCUSSION TITLE
How can I get every user who has exactly ROLE_USER

### DISCUSSION THREAD

**Discussion by AbdelilahDerfoufi** [symfony/symfony#45502] [Q&A]
I used querybuilder method to get all the users who has ROLE_USER as a role for an admin dashboard.

Here is the method I wrote :

`{
        $role = 'ROLE_USER';

        return $this->createQueryBuilder('u')
            ->where('u.roles = :role')
            ->setParameter('role', $role)
            ->getQuery()
            ->getResult();
    }
`

**Comment by chyshyya** [comment]
We add repository method:
```
    public function findMembers($role)
    {
        $qb = $this->createQueryBuilder('u');

        $qb
            ->andWhere($qb->expr()->like('u.roles', ':role'))
            ->setParameter('role', '%'.$role.'%')
            ->orderBy('u.username', 'ASC');

        return $qb
            ->getQuery()
            ->getResult();
    }
```

**Reply by chyshyya** [reply]
This is as example, because if role ROLE_USER and ROLE_USER_RICH, but you try to find ROLE_USER using our method it will find two roles together. Just a way to think. You can find with `->setParameter('role', '%"'.$role.'"%')` with `"`, as it saved in DB for example.
But if somebody knows better way, I'd like to see it and say thanks too.

**Reply by AbdelilahDerfoufi** [reply]
I tried it and I get nothing in my twig , check if I'm correct controller wise:

`#[Route('manager/users', name:'manager_users')]
    public function showUsers(UserRepository $userRepository)
    {
        return $this->render('manager/users.html.twig',[
            'users' => $userRepository->findMembers('ROLE_USER'),
        ]);
    }`

**Reply by chyshyya** [reply]
Did you add findMembers method to UserRepository?

**Reply by chyshyya** [reply]
I see... ROLE_USER added by default in symfony architecture to all users you created (check method getRoles() of User entity). You can use `$userRepository->findAll()` to return all of users stored in database. Look to your database and chech if field roles contain ROLE_USER string at all. I think answer is no.

**Reply by AbdelilahDerfoufi** [reply]
> Did you add findMembers method to UserRepository?

Yes.

**Reply by AbdelilahDerfoufi** [reply]
> I see... ROLE_USER added by default in symfony architecture to all users you created (check method getRoles() of User entity). You can use `$userRepository->findAll()` to return all of users stored in database. Look to your database and chech if field roles contain ROLE_USER string at all. I think answer is no.

Ah yes, you're right

**Reply by AbdelilahDerfoufi** [reply]
Then I should use not like ROLE _ADMIN and without getting ROLE_SUPER_ADMIN or something qnd I should get users who have RoLe_user and [] as roles, right?

**Reply by chyshyya** [reply]
I think that if you want to get users with ROLE_USER, but they must have only this role and not ROLE_CLIENT or ROLE_ADMIN or any other, then best choice to get them findMembers('[]') or to write other method that will check if roles equals '[]'

**Reply by AbdelilahDerfoufi** [reply]
I can add a condition if($role = '[]' || $role = 'ROLE_USER')
and we return qb, maybe this can work?

**Comment by endelwar** [comment]
I covered this topic in a blog post a while back: https://endelwar.it/2020/08/filter-users-by-role-in-symfony-5/
The solution still works with Symfony 6.
